Add more checks to custom Function

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33069 Add more checks to custom Function**
* #33068 Fix version counter bump in cpp Function

This PR adds the following:
- Warn when a non-input Tensor is given to `mark_dirty()` as it is not needed.
- Raise an error if we modify inplace an input that is a view and that we have multiple output. This setting is not handled by `CopySlices` and will raise a cryptic error during the backward.
- Raise an error if an input is modified inplace but not returned. That will prevent the graph rewrite from being done correctly.

Differential Revision: [D19791563](https://our.internmc.facebook.com/intern/diff/D19791563)

albanD committed:
Add more checks to custom Function

[ghstack-poisoned]

yf225 commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -92,9 +109,17 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+  // All the modified Tensors must be returned as is for the rewrite to be valid.
+  for (auto& dirty_input : dirty_inputs) {
+    TORCH_CHECK(outputs_impl.count(dirty_input) > 0,
+                "Some elements marked as dirty during the forward method where not returned as output. The"
```
nit:
```suggestion
                "Some elements marked as dirty during the forward method were not returned as output. The"
```

albanD replied:
Thanks, upated.

albanD committed:
Update on "Add more checks to custom Function"


This PR adds the following:
- Warn when a non-input Tensor is given to `mark_dirty()` as it is not needed.
- Raise an error if we modify inplace an input that is a view and that we have multiple output. This setting is not handled by `CopySlices` and will raise a cryptic error during the backward.
- Raise an error if an input is modified inplace but not returned. That will prevent the graph rewrite from being done correctly.

Differential Revision: [D19791563](https://our.internmc.facebook.com/intern/diff/D19791563)

[ghstack-poisoned]

albanD committed:
Update on "Add more checks to custom Function"


This PR adds the following:
- Warn when a non-input Tensor is given to `mark_dirty()` as it is not needed.
- Raise an error if we modify inplace an input that is a view and that we have multiple output. This setting is not handled by `CopySlices` and will raise a cryptic error during the backward.
- Raise an error if an input is modified inplace but not returned. That will prevent the graph rewrite from being done correctly.

Differential Revision: [D19791563](https://our.internmc.facebook.com/intern/diff/D19791563)

[ghstack-poisoned]
